lib/ansible/module_utils/oracle/oci_utils.py

```diff
@@ -0,0 +1,2007 @@
+
+    return result
+
+
+def _get_attributes_to_consider(exclude_attributes, model, module):
+    """
+    Determine the attributes to detect if an existing resource already matches the requested resource state
+    :param exclude_attributes: Attributes to not consider for matching
+    :param model: The model class used to create the Resource
+    :param module: An instance of AnsibleModule that contains user's desires around a resource's state
+    :return: A list of attributes that needs to be matched
+    """
+
+    # If a user explicitly requests us to match only against a set of resources (using 'key_by', use that as the list
+    # of attributes to consider for matching.
+    if "key_by" in module.params and module.params["key_by"] is not None:
+        attributes_to_consider = module.params["key_by"]
+    else:
+        # Consider all attributes except freeform_tags as freeform tags do not distinguish a resource.
+        attributes_to_consider = list(model.attribute_map)
+        if "freeform_tags" in attributes_to_consider:
+            attributes_to_consider.remove("freeform_tags")
+        # Temporarily removing node_count as the exisiting resource does not reflect it
+        if "node_count" in attributes_to_consider:
+            attributes_to_consider.remove("node_count")
+    _debug("attributes to consider: {0}".format(attributes_to_consider))
+    return attributes_to_consider
+
+
+def _is_resource_active(resource, dead_states):
+    if dead_states is None:
+        dead_states = DEAD_STATES
+
+    if "lifecycle_state" not in resource.attribute_map:
+        return True
+    return resource.lifecycle_state not in dead_states
+
+
+def is_attr_assigned_default(default_attribute_values, attr, assigned_value):
+    if not default_attribute_values:
+        return False
+
+    if attr in default_attribute_values:
+        default_val_for_attr = default_attribute_values.get(attr, None)
+        if isinstance(default_val_for_attr, dict):
+            # When default value for a resource's attribute is empty dictionary, check if the corresponding value of the
+            # existing resource's attribute is also empty.
+            if not default_val_for_attr:
+                return not assigned_value
+            # only compare keys that are in default_attribute_values[attr]
+            # this is to ensure forward compatibility when the API returns new keys that are not known during
+            # the time when the module author provided default values for the attribute
+            keys = {}
+            for k, v in iteritems(assigned_value.items()):
+                if k in default_val_for_attr:
+                    keys[k] = v
+
+            return default_val_for_attr == keys
+        # non-dict, normal comparison
+        return default_val_for_attr == assigned_value
+    else:
+        # module author has not provided a default value for attr
+        return True
+
+
+def has_user_provided_value_for_option(module, option):
```

Comment:
The aliases should be available via the standard name. I'm not sure what this function is for

ie
```python
display_name=dict(type="str", required=False, aliases=["name"]),
foo = module.params["display_name"]
# foo would be set to the what's passed in as `display_name` or `name`

```

Reply:
The function `has_user_provided_value_for_option()` figures out if an user has actually provided value for `option` using standard name or alias. This helps in preparing the list of attributes of a resource which need to be updated in the function `get_attr_to_update()`.

`module.params` has all the standard options set to some value(user provided, default or `None`) irrespective of whether user provided it or not.

Reply:
But *why* to do you need to figure out whether the option was provided via the main name or an alias? No other module makes any distinction by which name the option values are passed AFAIK. So why break with this convention?

Reply:
Thanks @felixfontein for reviewing this closely.

We have generalized the update utility[`check_and_update_resource()`] which is used by all the OCI Ansible modules for update scenarios.

To have a consistent design, the names of attributes of the update model of a resource become 'options with main name or alias' in the Ansible module for the resource.

For example, attributes: `defined_tags, display_name, freeform_tags` of the update model: https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/core/models/oci.core.models.UpdateVcnDetails.html are options in `oci_vcn` Ansible module and attribute:`display_name` could be alias for an option:`name`.

Then while iterating over the update model of a resource to figure out if the value of an attribute is actually provided by the user, we look for this attribute name in main names or aliases. If we don’t look in aliases, we may miss out as in the following case:

`display_name` could be alias for an option having standard name as `name`. While iterating on the update model to check if value of attribute:`display_name` is provided, we have to look for it in aliases too if user has used option:`name` to specify the value since `params["name"]` will be set and not `params["display_name"]`.

Reply:
If `name` is an alias of `display_name`, and the user sets `name: xxx`, `params['display_name']` will be `'xxx'`. Ansible does that internally so no module has to do that.

Reply:
Agree and it is true for `module.params`.

In `has_user_provided_value_for_option()`, we are processing return value of `_load_params()`, which is raw user input dictionary and has not gone through the default value substitution and alias processing. 

We need raw user input so that we can distinguish between cases when (a) a user explicitly assigns a value to an option and (b) a value gets assigned because of default value for the option.

Once we know what options are explicitly specified by a user, we update only those attributes.

Reply:
I think the standard way to do this (i.e. what all other modules do) is to not specify a default value for that option, and use `None` as "was not specified by user". I don't know how the core team sees this. If modules should be allowed to know where a value comes from (user, default value, environment, ...), I think there should be a common mechanism which can be used by all modules.

Reply:
@sivel @samdoran hey guys do you have any input on this?

Reply:
This feels a lot like reimplementing argument spec validation methods inside a module. I would advise against this.

> We need raw user input so that we can distinguish between cases when (a) a user explicitly assigns a value to an option and (b) a value gets assigned because of default value for the option.

I don't fully understand this need. The module should not care if a user provided the value or if it came from the default setting in the argument spec. It should only care about the value. Treating the value differently based on where it was set, particularly if the provided value is the same as the default value, seems overly complicated.

Reply:
Thanks @felixfontein and @samdoran for your inputs. 

We have a few module options in some OCI Ansible modules which are common to create and update operations. These options have defaults during create operation(as API puts in default value during create operation if nothing is specified) but don't have any defaults for update operation. 

As suggested under `options:`->`default:` on https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fields, we preferred mentioning default values for these options using the `default` construct instead of adding to option's documentation about defaults.

So during update operation when a user hasn't specified a value for such an option, the option gets substituted by the default value. Thus we have to check if the user provided the value or it came from the default setting in argument spec before we actually proceed for the update operation. If we don't make this check, we will end up updating the attribute of the resource even if the user didn't intend to.

Reply:
function removed